### PR TITLE
feat(bitmap): store multi bitmaps for different fragment in a SST

### PR DESCRIPTION
**Pull request by soundOfDestiny**
## What's changed and what's your intention?

store multi bitmaps for different fragment in a SST

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

### CODE REVIEW

File: src/storage/src/hummock/sstable/mod.rs [hunk 1]
```diff
@@ -125,11 +125,40 @@ impl BlockMeta {
...
+    #[inline]
+    pub fn encoded_size(&self) -> usize {
+        8 /* table_id + bitmap len */ + self.bitmap.len()
+    }
+}
+
 #[derive(Clone, PartialEq, Eq, Debug)]
 pub struct SstableMeta {
     pub block_metas: Vec<BlockMeta>,
+    pub bitmaps: Vec<Bitmap>,
```
Comment (xx01cyx) [inline src/storage/src/hummock/sstable/mod.rs:160]:
```suggestion
    pub vnode_bitmaps: Vec<Bitmap>,
```
Ditto.

File: src/storage/src/hummock/sstable/mod.rs [hunk 2]
```diff
@@ -144,8 +173,9 @@ impl SstableMeta {
     /// ```plain
     /// | N (4B) |
     /// | block meta 0 | ... | block meta N-1 |
+    /// | M (4B) |
+    /// | bitmap 0 | ... | bitmap M-1 |
```
Comment (xx01cyx) [inline src/storage/src/hummock/sstable/mod.rs:177]:
```suggestion
    /// | vnode bitmap 0 | ... | vnode bitmap M-1 |
```
Ditto.

File: src/storage/src/hummock/sstable/builder.rs [hunk 3]
```diff
@@ -65,8 +67,8 @@ pub struct SSTableBuilder {
     block_metas: Vec<BlockMeta>,
     /// Hashes of user keys.
     user_key_hashes: Vec<u32>,
-    /// Bitmap of value meta.
-    bitmap: [u8; VNODE_BITMAP_LEN],
+    /// `table_id` -> Bitmaps of value meta.
+    bitmaps: BTreeMap<u32, [u8; VNODE_BITMAP_LEN]>,
```
Comment (xx01cyx) [inline src/storage/src/hummock/sstable/builder.rs:71]:
```suggestion
    vnode_bitmaps: BTreeMap<u32, [u8; VNODE_BITMAP_LEN]>,
```

**Review by xx01cyx — COMMENTED** [review]
Rest LGTM.

File: src/storage/src/hummock/sstable/mod.rs [hunk 4]
```diff
@@ -125,11 +125,40 @@ impl BlockMeta {
     }
 }
 
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct Bitmap {
```
[the remark this answers is not in the document]
Comment (soundOfDestiny) [reply src/storage/src/hummock/sstable/mod.rs:129]: fixed

File: src/storage/src/hummock/sstable/mod.rs (hunk 2 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/sstable/mod.rs:177]: fixed

File: src/storage/src/hummock/sstable/builder.rs (hunk 3 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/sstable/builder.rs:71]: fixed

**Comment by hzxa21** [comment]
close?

**Comment by soundOfDestiny** [comment]
> close?

ok

**Pull request closed without merging**